### ISSUE TITLE
Read Only is omitted for named struct fields

### ISSUE THREAD

**Issue by c-thiel** [juhaku/utoipa#596]
Currently when using `toSchema` on a struct that has another struct as a named field, the `read_only` attribute is omitted in the resulting schema for this type. Here is a full example:

```rust
use serde::Serialize;
use utoipa::ToSchema;

#[derive(Serialize, ToSchema)]
pub struct Child {
    pub name: String,
    pub description: String,
}

#[derive(Serialize, ToSchema)]
pub struct Parent {
    pub name: String,
    #[schema(read_only)]
    pub description: String,
    #[schema(read_only)]
    pub child: Child,
}

fn main() {
    let schema = Parent::schema();
    println!("{}", serde_json::to_string_pretty(&schema).unwrap());
}
```

This prints:
```json
[
  "Parent",
  {
    "type": "object",
    "required": [
      "name",
      "description",
      "child"
    ],
    "properties": {
      "name": {
        "type": "string"
      },
      "description": {
        "type": "string",
        "readOnly": true
      },
      "child": {
        "$ref": "#/components/schemas/Child"
      }
    }
  }
]
```

Note that the child has no readOnly Attribute.
The correct way to render this would be for OAS 3.0.X:

```json
"child": {
    "readOnly": true,
    "allOf": [
         "$ref": "#/components/schemas/Child"
    ]
}
```

This actually is supported by Swagger - the field does not show up for POST methods.
However in the `Schemas` section of the webpage, this seems to not be rendered.

In OAS 3.1 I believe that sibling keywords are supported out of the box.

Do you agree on this? Thanks for some feedback :)

**Comment by juhaku** [comment]
> Do you agree on this? Thanks for some feedback :)

That is correct, (assuming of course that the allOf hack is supported for the `readOnly` property.) That is the OpenAPI 3.0.x spec does not support multiple types at object nor it is a full JSON Schema spec compatible. Moreover the `$ref` type that is being used for referencing another schema does not have any other fields in OpenAPI 3.0.x spec thus there is a need for the `allOf` hack. 

The `allOf` hack is used for some type, was it currently for `nullability` but not for all properties indeed. I don't remember what properties the `$ref` supported in OpenAPI 3.1.x spec but the support was more throughout than in 3.0.x.

Yet there was a request to support description on reference fields and that support would need this `allOf` hack to in place for references as well but it was never implemented. Perhaps this could be added in future probably some where around when the openapi 3.1.x spec support will be added.

**Comment by c-thiel** [comment]
So enabling that hack for 3.0.X optionally is not a good idea?

If so, then +1 for 3.1 :)

**Comment by juhaku** [comment]
I would say, it could be, but just need to keep an eye for the 3.1 support so that it would not intorduce much changes that will be counter intuitive for 3.1. Also at the moment I am no able to focus on this subject nevertheless this could be looked into as one of the next few things. 

Generally this would need some changes on how `$ref`s are now being processed. My idea was to have by default nearly full support for 3.0 spec and then with feature flag enabling support for 3.1 eventually. But perhaps if necessary the `allOf` hack could be also enabled via feature flag if it comes to that.

**Comment by Drifter-J** [comment]
Hi folks, I raised a [PR](https://github.com/juhaku/utoipa/pull/1550/) for covering this case (for both read_only and write_only) @juhaku @c-thiel PTAL!
